### DISCUSSION TITLE
3D-model Stage Interface Discussions

### DISCUSSION THREAD

**Discussion by SuperFromND** [ikemen-engine/Ikemen-GO#1328] [Engine development]
This thread is for discussing the interface(s) for a currrently-hypothetical implementation of fully 3D, model-based stages in IKEMEN. This is continued from a brief conversation in the IKEMEN Discord.
**NOTE: As of this writing these are merely discussions, no work has started on this to my knowledge and I wouldn't expect it to for quite a while!**

## File Formats 
For an initial implementation, I think it'd make sense to stick with only one format, preferrably:
- a single-file format for convenience when sharing stages (i.e. not having to share dozens of texture files along with every stage )
- something supported by lots of programs (or at the very least Blender, since Open Source:tm:)
- a format that can contain all of its necessary data (mesh, textures, lighting, possibly animation data)

![Screenshot of Blender showing tons of different formats it can export to](https://github.com/ikemen-engine/Ikemen-GO/assets/22881403/fc4a8966-4f38-4c19-9ca4-88b65c169efe)
It turns out there are a _LOT_ of standards for 3D model files. However, in my personal experience, it seems like [the FBX format](https://en.wikipedia.org/wiki/FBX) ticks most if not all of these boxes, and is quite common and well-understood by 3D modellers for other games that allow custom 3D content, such as VRchat. 

For a more open and newer (but less commonly supported) format, [glTF v2.0](https://www.khronos.org/gltf/) may be worth investigating as well as it can also tick most of these boxes via the `.gltf` (embedded) format.

### Libraries
As a bit of an aside, I looked a tiny bit into Go's library ecosystem to see if someone wrote any parsers for 3D object files already. Turns out there's a couple:
- [oakmound/ofbx](https://github.com/oakmound/ofbx) supports FBX parsing, but is claimed in its readme as "currently at version 0.1 and has limited functionality". Has not been updated since 2021. MIT-licensed.
- [udhos/gwob](https://github.com/udhos/gwob) supports parsing only Wavefront .obj/mtl files; consisting purely of mesh data and texture coordinates in the first file and texture/material data in the second; no animation data sadly. MIT-licensed.
- [qmuntal/gltf](https://github.com/qmuntal/gltf) contains both an encoder and a decoder for glTF v2.0 files. Explicitly claims to aim for production-ready stability, a good fit for an engine like ours. BSD-2 licensed.

## Interface Proposal
The _interface_ (no relation to GUIs or screenpacks or anything like that) refers to how end-users (specifically stage authors and editors) can tell IKEMEN to use a 3D model in stages. In other words, it's how the DEF file is written out.

Below is an example of how IKEMEN currently handles 2D sprites for stages.
```ini
[BGdef]
spr = stage0.sff

[BG 0]
type  = normal
spriteno = 0, 1
start = 0, 185
tile = 1, 0
```
This section declares a `BGdef` and loads a 2D sprite file, `stage0.sff`, and then defines a `BG` element to display it. There's a lot of additional parameters that could go there of course, but I stripped it down to the basics for simplicity sake.

To differentiate 3D stages from 2D ones (and to no-doubt make parsing easier), My first proposal consists of introducing two new element labels: `3DBGdef` and `3DBG`.

First, `3DBGdef` specifies the model file in question to load. There should only be one per stage; below are three different versions depending on which format(s?) we end up supporting.
```ini
# if we're using FBX
[3DBGdef]
model = Example.fbx

# if we're using glTF
[3DBGdef]
model = Example.glft

# if we're using OBJ/MTL
[3DBGdef]
obj = Example.obj
mtl = Example.mtl
```

Then, `3DBG` defines the model's location, scale, rotation, and so forth. It's hard to really make too many parameters here without deciding a format to use first (hence the lack of indexes for example; we're using the entirety of the mesh here!), but this should give you a basic idea of how it would work.

```ini
[3DBG My Cool Platform]
; X, Y and Z coordinates are used here
; There's conflicting standards on which axis represents what, but in this case let's assume
; X = horizontal, Y = vertical, and Z = depth forwards/backwards from the screen
; this would make the most sense to those used to MUGEN coordinates

; [0, 0] is understood to mean the *center/origin* of the stage in a 3D context, see below
; Worth noting that models have their own position offsets (origins) built-in, just like MUGEN sprites do
; These are all intended to be float values
start = 0, 0, 0
rotation = 0, 0, 0
scale = 1, 1, 1
```

### Materials
Materials (how a 3D object is shaded and textured) are most likely already handled by whatever format we end up choosing. If for whatever reason we don't have a material (or the author just wants to use a uniform value different from the original mesh), I think that could potentially be exposed as parameters too.

For those who aren't 3D modellers, general consensus with just about every 3D model format is to use at least three values; these have different names depending on the standard and program, but going by Blender naming they are:
- **Color map** a.k.a **Base color** a.k.a **Albedo map**: The base color or texture of the object. If specified via state controller, it should use a solid color across the entire mesh, overwriting any sort of texture map or other solid color that was originally there. Values 0-255 for each value, representing RGB.
- **Roughness map** a.k.a **Diffuse map**: Defines how "rough" a given surface is, with a value of 0 meaning a plastic-looking glossy surface and a value of 1 meaning matte/rough. Usually defaults to 0.5.
- **Specular map** a.k.a **Reflective map**: Defines how "shiny" a given surface is (the white highlights from a light), a value of 0 meaning no shine and a value of 1 meaning maximum shine. Usually defaults to 0.5.
```ini
[3DBG Let's Set Some Overrides]
; These values are completely optional and in fact should be *ignored* unless
; explicitly present and set like this. These values are normally provided as texture maps 
; by the model file itself. By contrast these can only use single uniform values
; that apply to the entirety of the mesh.

; Anyways, these values would create an extremely shiny, bright-red mesh.
color = 255, 0, 0
roughness = 0
specular = 1
```

### Coordinate Space
Now, there's a bit of a unique problem here. See, MUGEN specifications (and IKEMEN by extension) use pixel coordinates to position everything. X and Y coordinates are usually measured in the hundreds of pixels, in integer format like so:
```ini
start = 0, 185
```

However, 3D models by their very nature use *much* smaller float values, rarely exceeding `30.0f` on any given coordinate.
```obj
# Vertexes of a cube in Wavefront OBJ format
v 1.000000 1.000000 -1.000000
v 1.000000 -1.000000 -1.000000
v 1.000000 1.000000 1.000000
v 1.000000 -1.000000 1.000000
v -1.000000 1.000000 -1.000000
v -1.000000 -1.000000 -1.000000
v -1.000000 1.000000 1.000000
v -1.000000 -1.000000 1.000000
```

Therefore, I think that when using models an internal multiplication of 100 should be applied to a given model's coordinates/rotation/scale, which should bring most models more in line with MUGEN's pixel coordinate values.

There's also the matter of how `localcoord` should factor into these models. Personally, I think it's best to simply ignore `localcoord` entirely as it doesn't really make much sense for 3D stuff, but I'm curious for feedback on this part.

### Fighter Positioning
I think that the fighters should be aligned to Z-coordinate 0, and their behavior shouldn't really change at all from being in a 3D stage. The effect would look similar to _Marvel vs. Capcom 2_; 2D sprited characters moving and behaving on a 2D plane that just happens to be projected onto a 3D world.
![Diagram showing what this would look like in 3D space](https://github.com/ikemen-engine/Ikemen-GO/assets/22881403/4ffe0124-c776-41fa-a208-183113e7a259)

### HUDs and Explods
I haven't fully thought through how HUDs should be handled in this context. Maybe the HUD should just be drawn after the stage model is, effectively overlaying it on top no matter what.

Explods are a much trickier matter. Should they be rendered at Z=0 just like characters? But then how do you handle layering with any 3D stages that have geometry that covers up Z>0? Maybe determine where to render them based on their `sprpriority` value?

## Other
As you can probably tell, I got a *little* sidetracked while writing what was originally a post maybe five lines long at most. I probably forgot something while writing this monster, so please don't hesitate to ask if I need to clarify anything further!

**Comment by ghost** [comment]
With this we can make true 3D stages and give a Marvel vs Capcom 2 feel, I really like that game's unique graphical style, 2d characters in smooth, animated fully 3d stages.

**Comment by SuperFromND** [comment]
Eiton's been working on implementing this for the past couple days, and I was asked to catalog some known issues here in this discussion for organization's sake, so:

- Objects with no material will crash the game with a reference to a nil pointer, I found this when I wanted to test animation with a simple cube, and it took me a bit to realize it wasn't my attempt to animate, it was that adding a cube doesn't automatically give it a default material.

- An object with only 1 animation keyframe will cause the game to freeze on it's first frame, this is because each object will animate on it's own timeline, so if an object has 1 keyframe, I assume it tries to loop through that keyframe, but it has a transition of 0 seconds, so the game gets stuck trying to animate that.

(Thank you to Orange Dolphin on Discord for finding these issues!)

**Reply by Orange-Dolphin** [reply]
A follow up discovery, with regular stages in Ikemen, the stage is reset when you return to the menu and start a new match. With 3d models, the animation does not reset, so wherever the animation ended when you go to the menu is where the animation starts in the next match. I have attached a video for reference of this, the 2d stage is reset, as such it is expected that the Clocktown stage should always start with the hour wheel in the same position, but after the match restart, it starts 2 hours ahead, where the first match ended.

This may create issues with any sprite animations that are timed to something in the 3d animation, as well as with the Clocktown stage there is an attached character who plays the clock turning sound, but due to the animation not resetting, it desyncs the timing after the first match.
https://github.com/ikemen-engine/Ikemen-GO/assets/147012366/7d725e61-ad4f-46b2-a522-48e35c0855e0

**Comment by LasombraDemon** [comment]
>there is an attached character who plays the clock turning sound, but due to the animation not resetting, it desyncs the timing after the first match. https://github.com/ikemen-engine/Ikemen-GO/assets/147012366/7d725e61-ad4f-46b2-a522-48e35c0855e0

^ That can be solved using GameTime instead of Time when coding the AttachedChar. But yeah, they *should* reset.

My two cents as a stage dev: stage elements *need* to be able to be addressed separately, to be managed by the engine code, via BGCtrls and IDs. If I understand correctly, the whole 3D element is a single thing, right now... And that's a big problem. A 3D barrel in front of the players should be adressable separately from the huge 3D mountain in the background.

This is because BGCtrls are the key to truly complex stage behaviour that allows for more advanced things: I'd build the format so that "3DBGCtrls" are possible in the future.

I know it is still too early for this, but I'd take extra care to make things expandable up to *at least* the preexisting BGCtrls for sprites.

**Comment by SuperFromND** [comment]
> I know it is still too early for this, but I'd take extra care to make things expandable up to _at least_ the preexisting BGCtrls for sprites.

This is a completely fair point and something I didn't really account for initially as I don't really mess with AttachedChars and the like (plus I wanted the implementation to be as simple as possible for newcomers, as well as to make it easier on the programming side of things).

So since we've pretty much settled on the glTF standard, we can actually design with the capabilities of that format in mind; namely, glTF files are capable of storing the details of individual mesh objects according to [their spec sheet](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#reference-mesh):

![image](https://github.com/ikemen-engine/Ikemen-GO/assets/22881403/d877fd83-338e-47db-b524-d69c422dfd02)

Note the `name` field there, which if I'm parsing this spec sheet right, is equivalent to how objects can be named in programs like Blender:

![image](https://github.com/ikemen-engine/Ikemen-GO/assets/22881403/14c3ad56-5f9e-47a7-94e8-1415a3b1252c)
![image](https://github.com/ikemen-engine/Ikemen-GO/assets/22881403/39b029bd-bcd6-43da-86bc-8730933e2271)

So with that in mind, perhaps we can amend how `[Model]` elements are defined to something like this:

```ini
[Model]
; new parameter: "name"
; This parameter takes the name of a given mesh object as defined in the glTF file.
; When specified, only that particular object is rendered within that block, similar to BG elements for 2D stages.
; If this parameter is left blank or undefined, then default to loading the entire scene like how we currently do.
name = "House"

; Parameters currently available in Eiton's fork

; Offset is still understood here to mean the origin of the world {0, 0, 0}. 
; Since objects can have their own pivot point, the pivot should be taken into account when implementing this.
; (In other words: the object's pivot should be at 0, 0, 0 if this offset value is also set to 0, 0, 0.)

offset = 0, 0, 0
scale = 1, 1, 1

; Not implemented (at least not yet), but included for the sake of completeness
rotation = 0, 0, 0

; From skimming the docs, this parameter is apparently essential for interactive stages to work.
; It's a user-set numeric ID that can be used to point to a specific BG element.
sctrlid = 1
```

I'm admittedly not very familiar with exactly *how* interactive stages work just yet, but from skimming the docs this should allow you to at least reference specific parts of a 3D stage under this format.

**Comment by LasombraDemon** [comment]
This is an excellent idea! :D As much as I appreciate the SCTRLID (which is used for AttachedChars), an ID reference is also required (which is a normal, MUGEN stage feature implementation). Those allow to change things based on a timer alone, not player interaction.

This section shows how BGCtrls work: https://www.elecbyte.com/mugendocs/bgs.html#background-controllers

And this section explains it a bit more: https://mugenguild.com/forum/topics/background-controllers-stages-170304.0.html

These should be considered first because we can *then* hit the actual stage elements via AttachedChar's ModifyBGCtrl. Without the BGCtrls, we cannot affect stage stuff via AttachedChar.

**Reply by SuperFromND** [reply]
Ah yeah, that ID thing slipped my mind. So a regular `id` parameter in addition to an `sctrlid` for these blocks, right?

```ini
; From skimming the docs, this parameter is apparently essential for interactive stages to work.
; It's a user-set numeric ID that can be used to point to a specific BG element.
sctrlid = 1

; Carry-over from regular M.U.G.E.N stage element specifications
; Also important for interactive stages as well as interfacing with background controllers.
id = 6
```

**Reply by LasombraDemon** [reply]
Indeed! That way we can then think how *stages* can affect them. And then, how *AttachedChars* can affect them.
